massa-protocol-worker/src/protocol_worker.rs

```diff
@@ -226,27 +226,42 @@ mod nodeinfo {
             self.known_operations.contains(op)
         }
 
-        /// insert a block in wanted list of a node.
-        /// Note that it also insert the block as a not known block for that node.
-        pub fn insert_wanted_blocks(
-            &mut self,
-            block_id: BlockId,
-            max_node_wanted_blocks_size: usize,
-        ) {
-            self.wanted_blocks.insert(block_id, Instant::now());
-            while self.known_blocks.len() > max_node_wanted_blocks_size {
+        /// Remove the oldest items from known_blocks
+        /// to ensure it contains at most max_node_wanted_blocks_size items.
+        /// This algorithm is optimized for cases where there are no more than a couple excess items, ideally just one.
+        fn remove_excess_wanted_blocks(&mut self, max_node_wanted_blocks_size: usize) {
+            while self.wanted_blocks.len() > max_node_wanted_blocks_size {
                 // remove oldest item
                 let (&h, _) = self
-                    .known_blocks
+                    .wanted_blocks
                     .iter()
                     .min_by_key(|(h, t)| (*t, *h))
                     .unwrap(); // never None because is the collection is empty, while loop isn't executed.
-                self.known_blocks.remove(&h);
+                self.wanted_blocks.remove(&h);
             }
         }
 
-        /// If given node previously manifested it wanted given block.
-        pub fn contains_wanted_block(&mut self, block_id: &BlockId) -> bool {
+        /// Insert a block in the wanted list of a node.
+        /// Also lists the block as not known by the node
+        pub fn insert_wanted_block(
+            &mut self,
+            block_id: BlockId,
+            max_node_wanted_blocks_size: usize,
+            max_node_known_blocks_size: usize,
+        ) {
+            // Insert into known_blocks
+            let now = Instant::now();
+            self.wanted_blocks.insert(block_id, now);
+            self.remove_excess_wanted_blocks(max_node_wanted_blocks_size);
+
+            // If the node wants a block, it means that it doesn't have it.
+            // To avoid asking the node for this block in the meantime,
+            // mark the node as not knowing the block.
+            self.insert_known_blocks(&[block_id], false, now, max_node_known_blocks_size);
```

Comment:
You don't prune known blocks ?

Reply:
the insert_known_blocks function does

Reply:
here I also separated remove_excess_known_blocks for more clarity